libsolidity/analysis/TypeChecker.cpp

```diff
@@ -1622,6 +1622,24 @@ bool TypeChecker::visit(MemberAccess const& _memberAccess)
 	}
 	else if (exprType->category() == Type::Category::FixedBytes)
 		annotation.isLValue = false;
+	else if (exprType->category() == Type::Category::Contract)
+	{
+		if (auto callType = dynamic_cast<FunctionType const*>(type(_memberAccess).get()))
+		{
+			auto kind = callType->kind();
+			auto contractType = dynamic_cast<ContractType const*>(exprType.get());
+			solAssert(!!contractType, "Should be contract type.");
+
+			if (
+				(kind == FunctionType::Kind::Send || kind == FunctionType::Kind::Transfer) &&
+				!contractType->isPayable()
+			)
+				m_errorReporter.typeError(
+					_memberAccess.location(),
+					"Value transfer to a contract with a non-payable fallback function."
+				);
+		}
+	}
 	else if (TypeType const* typeType = dynamic_cast<decltype(typeType)>(exprType.get()))
```

Comment:
Double check this else here. I fear there might be some overlap with the code I just added, but I couldn't reproduce it.